### PR TITLE
tests/posix/common/src/clock.c: Dummy change to trigger issue with m2gl025_miv board in CI

### PR DESCRIPTION
**Pull request by pfalcon**
…gl025_miv board in CI.

Backing issue: #25017

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

### CODE REVIEW

**Comment by pfalcon** [comment]
For reference, current m2gl025_miv failures are:

* https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/71605/1/tests
* https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/71605/3/tests

**Comment by pfalcon** [comment]
https://github.com/zephyrproject-rtos/zephyr/issues/25017 is created to track issue exposed by this patch.

**Comment by pfalcon** [comment]
To quote, the actual failure looks like:
~~~
20:30:06.1773 [DEBUG] uart: [+0.85ms host +0s virt 0.7s virt from start]   starting test - test_posix_realtime
20:30:06.1781 [DEBUG] uart: [+0.78ms host +0s virt 0.7s virt from start]   POSIX clock set APIs
20:30:06.1790 [DEBUG] uart: [+0.95ms host +0.1ms virt 0.7s virt from start]   
20:30:06.1834 [DEBUG] uart: [+4.35ms host +0s virt 0.7s virt from start]       Assertion failed at WEST_TOPDIR/zephyr/tests/posix/common/src/clock.c:66: test_posix_realtime: (rts.tv_sec not equal to mts.tv_sec)
20:30:06.1839 [DEBUG] uart: [+0.54ms host +0s virt 0.7s virt from start]   Seconds not equal
20:30:06.1852 [DEBUG] uart: [+1.32ms host +0.1ms virt 0.7s virt from start]   FAIL - test_posix_realtime
~~~

**Comment by carlescufi** [comment]
@andyross is this something that could be related to recent timeout changes?

**Comment by pfalcon** [comment]
I added a commit which shows exact values in the failed comparison:

> 10:03:00.1941 [DEBUG] uart: [+0.67ms host +0s virt 0.7s virt from start]   starting test - test_posix_realtime
10:03:00.1947 [DEBUG] uart: [+0.61ms host +0s virt 0.7s virt from start]   POSIX clock set APIs
10:03:00.1962 [DEBUG] uart: [+1.5ms host +0.1ms virt 0.7s virt from start]   secs: mono: 69, real: 70
10:03:00.1964 [DEBUG] uart: [+0.21ms host +0s virt 0.7s virt from start]   
10:03:00.1988 [DEBUG] uart: [+2.37ms host +0s virt 0.7s virt from start]       Assertion failed at WEST_TOPDIR/zephyr/tests/posix/common/src/clock.c:67: test_posix_realtime: (rts.tv_sec not equal to mts.tv_sec)
10:03:00.1995 [DEBUG] uart: [+0.71ms host +0.1ms virt 0.7s virt from start]   Seconds not equal
10:03:00.2004 [DEBUG] uart: [+0.83ms host +0s virt 0.7s virt from start]   FAIL - test_posix_realtime

In other words, for some reason CLOCK_MONOTONIC has value of 69 seconds, while CLOCK_REALTIME - 70 seconds. (For a moment I thought that maybe this board has RTC and CLOCK_REALTIME has a real wallclock time, but nope, just a weird 1s diff).

**Comment by andyross** [comment]
Seems not impossible that this is tickled by the timeout changes, yeah.  Is tests/posix/common the only affected test?   Does anyone have a clue how posix implements MONOTONIC vs. REALTIME?  Those aren't kernel features (the difference is in how it treats changes to the system clock due to calibration, etc... -- things we don't have an API for).

FWIW: I see these in the kconfig for that board:

```
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=660000
CONFIG_SYS_CLOCK_TICKS_PER_SEC=100
```

Is 660 kHz **really** the right clock speed for the RISC-V machine timer on the board?  That seems odd.  The intent of that feature is that it's a high speed 64 bit counter running at or near CPU speed, like the x86 TSC, right?  Maybe it's supposed to e 66 MHz and someone slipped a few zeroes?

Also a tick rate of 100 Hz is silly on a hardware device.  We still do that in emulation due to host timing interactions, but tickless devices with hardware clocks should be much higher.  We use 10 kHz for platforms with base-10-divisible clock rates, usually.

**Comment by andyross** [comment]
Yeah, here's the only place in the code where we treat those differently:

https://github.com/zephyrproject-rtos/zephyr/blob/master/lib/posix/clock.c#L26

Basically REALTIME has a fixed offset from system startup (to reflect the "actual current time" from an external source, I guess), where MONOTONIC doesn't.and just returns a raw uptime.  If they're skewed, it's because something is setting the time offset to something other than zero.

But... that's not wrong, is it?  My guess is that if the test is failing it's because the test is making assumptions that aren't correct.

Alternatively the test has some code to set the time and **check** the offset?  But I don't see how that can fail looking at how simple that implementation is.

Broadly: I think the timeout changes are off the hook, and I'm pretty sure the POSIX layer looks fine too.  My guess is this is a test bug.

**Comment by pfalcon** [comment]
> Yeah, here's the only place in the code where we treat those differently: https://github.com/zephyrproject-rtos/zephyr/blob/master/lib/posix/clock.c#L26

Yeah, and in the latest CI build I even put printf debugging to see if something trashes memory in such a peculiar way that rt_clock_base.tv_sec gets exactly 1:

https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/71679/5/tests

~~~
11:32:59.1990 [DEBUG] uart: [+3.14ms host +0.1ms virt 0.7s virt from start]   rt_clock_base.tv_sec: 0, .tv_nsec: 0
11:32:59.2000 [DEBUG] uart: [+0.96ms host +0s virt 0.7s virt from start]   secs: mono: 69, real: 70
~~~

So, nope.

A couple of wild ideas:

1. The system is so deterministic that across half-dozen runs already made, the second bump happens exactly between 2 clock_gettime() calls. That already can give a smile, but a real LOL would be:
2. There's a bug in time arithmetics which happens around 69/70th second ;-).

Jokes aside, I'd still like to hear from folks who put that board into CI loop re: how much they tested the stuff before doing it. And apparently, from the board maintainers.

Neither of that is blocking so far of course, it would be a nice idea to run that on different real hw (of course, it doesn't fail in qemu). So far, didn't get to it.

And of course, we can patch stuff on the test side, and likely that's how it will end. It's just a pretty interesting case when "improvement" to CI loop actually becomes a source of instability, and if we don't treat those seriously, but silently work around, there won't be much good IMHO.

**Comment by andyross** [comment]
OK, for sure it's a buggy test, though I don't have a board and lack @pfalcon's dedication to testing this by firing stuff blindly into CI.  Here's the code that's failing:

https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/posix/common/src/clock.c#L58

It looks simple enough: it calls clock_gettime() twice, once with the REALTIME clock and once with MONOTONIC, and validates that the results are identical.

**But that's wrong**, because time passed between the two calls.  Clearly they aren't always going to be equal, they happened at different times!  In fact if you run exactly that sequence on x86 Linux (anything with a vdso-aware C library, so any distro of the past decade or so), it will fail every time, because the clock is provided by the TSC and that's sensitive to number of cycles that pass between the two calls.

Zephyr isn't that precise (yet), so in principle this could be made to work like it does on similar tests by ensuring tick alignment and guessing that two function calls won't take longer than a tick.  We could try this as a workaround:

```diff
diff --git a/tests/posix/common/src/clock.c b/tests/posix/common/src/clock.c
index 8465b7bcae..b420522514 100644
--- a/tests/posix/common/src/clock.c
+++ b/tests/posix/common/src/clock.c
@@ -55,6 +55,9 @@ void test_posix_realtime(void)
        struct timespec rts, mts;
        struct timeval tv;
 
+       /* Minimal sleep to align us to the next tick interval */
+       k_usleep(1);
+
        printk("POSIX clock set APIs\n");
        ret = clock_gettime(CLOCK_MONOTONIC, &mts);
        zassert_equal(ret, 0, "Fail to get monotonic clock");
```

**Comment by pfalcon** [comment]
> But that's wrong

But of course, even the test itself knows that and invites to be victimized: https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/posix/common/src/clock.c#L51 . And yet so far it didn't fail consistently.

> Zephyr isn't that precise (yet), so in principle this could be made to work like it does on similar tests by ensuring tick alignment and guessing that two function calls won't take longer than a tick.

That all doesn't explain why with 100% probability on a particular board 2 function calls "fall on a tick which increment the **seconds** counter". As that cannot happen, there's something wrong with that board, potentially other boards, and potentially other code.

 > +       k_usleep(1);

I'll try it on next iteration playing with this.

**Comment by pfalcon** [comment]
@andyross: Your hint with k_usleep(1) worked, thanks! So, I'm adding that workaround ahead of removing that test chunk.

**Comment by pfalcon** [comment]
Fix merged into https://github.com/zephyrproject-rtos/zephyr/pull/24964. Closing.

**Comment by andyross** [comment]
OK, cool.  Yeah, the problem with many hardware platforms is that the timer code is scary deterministic.  You get something that should "never" happen and it turns out to be 100% reliable because the reset behavior is cycle-exact.
